providers/provider_default_test.go

```diff
@@ -19,3 +23,35 @@ func TestRefresh(t *testing.T) {
 	assert.Equal(t, false, refreshed)
 	assert.Equal(t, nil, err)
 }
+
+func TestCreateSessionStateFromBearerToken(t *testing.T) {
+	minimalIDToken := jwt.StandardClaims{
+		Audience:  "asdf1234",
+		ExpiresAt: time.Now().Add(time.Duration(5) * time.Minute).Unix(),
+		Id:        "id-some-id",
+		IssuedAt:  time.Now().Unix(),
+		Issuer:    "https://issuer.example.com",
+		NotBefore: 0,
+		Subject:   "123456789",
+	}
+	// From oidc_test.go
+	verifier := oidc.NewVerifier(
+		"https://issuer.example.com",
+		fakeKeySetStub{},
+		&oidc.Config{ClientID: "asdf1234"},
+	)
+
+	key, _ := rsa.GenerateKey(rand.Reader, 2048)
+	rawIDToken, _ := jwt.NewWithClaims(jwt.SigningMethodRS256, minimalIDToken).SignedString(key)
+	idToken, err := verifier.Verify(context.Background(), rawIDToken)
+
+	session, err := (*ProviderData)(nil).CreateSessionStateFromBearerToken(context.Background(), rawIDToken, idToken)
+
+	assert.Equal(t, nil, err)
+	assert.Equal(t, rawIDToken, session.AccessToken)
+	assert.Equal(t, rawIDToken, session.IDToken)
+	assert.Equal(t, "", session.RefreshToken)
+	assert.Equal(t, "123456789", session.Email)
```

Comment:
What is the overall architectural rationale for forcing a `session.Email` to a `claim.Subject` when an `email` claim is missing? And not using the `claim.Subject` at all in the session and setting both the `session.Email` and `session.User` to the Email?

https://github.com/oauth2-proxy/oauth2-proxy/blob/f7b28cb1d362f54d13379136c37f0edc1303a1b0/providers/provider_default.go#L155

I left the existing logic untouched -- but shouldn't `session.User` == `claim.Subject` (That's what my OIDC provider results in going the normal route).

https://github.com/oauth2-proxy/oauth2-proxy/blob/f7b28cb1d362f54d13379136c37f0edc1303a1b0/providers/provider_default.go#L165

And should we allow `session.Email` to be "" if the IDToken is missing the claim?  And adjust the email domains validator to accept it if its set to `*` (if set to anything else it would fail since it is missing a `@` character)

Reply:
User ID claim was added so that we could make sure that a particular claim was added, to be considered the user, but the work hasn't been completed so it's only half way through

As for falling back, I'm not sure why we are doing that, i think we currently require all sessions to have a user and email field that's not empty, so I guess that's why? I think I need to do some archaeology on this